src/execution/riscv64/simulator-riscv64.h

```diff
@@ -110,44 +110,18 @@ static inline bool isSnan(double fp) { return !QUIET_BIT_D(fp); }
 #undef QUIET_BIT_D
 
 inline uint64_t mulhu(uint64_t a, uint64_t b) {
-  if (is_int32((int64_t)a) && is_int32((int64_t)b)) {
-    return ((((((a & 0xfffffffff) << 32) >> 32) *
-              (((b & 0xfffffffff) << 32) >> 32)) >>
-             32));
-  }
-  uint64_t t;
-  uint32_t y1, y2, y3;
-  uint64_t a0 = (uint32_t)a, a1 = a >> 32;
-  uint64_t b0 = (uint32_t)b, b1 = b >> 32;
-  t = a1 * b0 + ((a0 * b0) >> 32);
-  y1 = (uint32_t)t;
-  y2 = t >> 32;
-
-  t = a0 * b1 + y1;
-  y1 = (uint32_t)t;
-
-  t = a1 * b1 + y2 + (t >> 32);
-  y2 = (uint32_t)t;
-  y3 = t >> 32;
-
-  return ((uint64_t)y3 << 32) | y2;
+  __uint128_t full_result = ((__uint128_t)a) * ((__uint128_t)b);
+  return full_result >> 64;
 }
 
 inline int64_t mulh(int64_t a, int64_t b) {
-  if (is_int32(a) && is_int32(b)) {
-    return ((((((a & 0xfffffffff) << 32) >> 32) *
-              (((b & 0xfffffffff) << 32) >> 32)) >>
-             32));
-  }
-  int negate = (a < 0) != (b < 0);
-  uint64_t res = mulhu(a < 0 ? -a : a, b < 0 ? -b : b);
-  return negate ? ~res + (a * b == 0) : res;
+  __int128_t full_result = ((__int128_t)a) * ((__int128_t)b);
+  return full_result >> 64;
 }
 
 inline int64_t mulhsu(int64_t a, uint64_t b) {
-  int negate = a < 0;
-  uint64_t res = mulhu(a < 0 ? -a : a, b);
-  return negate ? ~res + (a * b == 0) : res;
+  __int128_t full_result = ((__int128_t)a) * ((__uint128_t)b);
```

Comment:
Is this logic correct? This looks the same as `mulh`. The original logic in these is coming from the Spike simulator, which should be well-tested - https://github.com/riscv/riscv-isa-sim

Reply:
The helper routines were wrong, so I wanted to decouple these to fix the issues one at a time.  If mulhu was correct then they're equivalent, it's just less code this way and the code that's there relies on the compiler to get the types correct.

Reply:
Ok,sorry,  I misread this before. It looks correct to me.